src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/Orleans3CompatibleStringKeyHasher.cs

```diff
@@ -43,28 +43,24 @@ public int Hash(byte[] data)
 
             var extendedLength = data.Length + 8;
 
-            var buffer = extendedLength switch
-            {
-                <= 32 => stackalloc byte[32],
-                <= 64 => stackalloc byte[64],
-                <= 128 => stackalloc byte[128],
-                <= 256 => stackalloc byte[256],
-                _ => Span<byte>.Empty
-            };
-
-            byte[] rentedBuffer = null;
-            if (buffer.IsEmpty)
-            {
-                // assuming code below never throws, so calling ArrayPool.Return without try/finally block for JIT optimization
-                rentedBuffer = ArrayPool<byte>.Shared.Rent(extendedLength);
-                buffer = rentedBuffer.AsSpan();
-            }
+            const int maxOnStack = 256;
+            var rentedBuffer = extendedLength > maxOnStack
+                ? ArrayPool<byte>.Shared.Rent(extendedLength)
+                : null;
+
+            // assuming code below never throws, so calling ArrayPool.Return without try/finally block for JIT optimization
+
+            var buffer = rentedBuffer is not null
```

Comment:
Why don't you write it as suggested (i.e. like use almost everywhere with .NET)?

Reply:
Correct me if I'm wrong.

When we call `ArrayPool.Rent` and don't call `ArrayPool.Return`, the buffer we used will be collected by GC. Even if GC will return the array to pool, nobody nows when the GC kicks in. So, further calls to `ArrayPool.Return` will allocate new arrays (we will quickly deplete the pool by frequent `ArrayPool.Rent` calls) which will lead to high memory traffic. That's why I return the array explicitly as soon as possible to make it recycled by further `ArrayPool.Rent` calls.

Reply:
small repro

```c#
[MemoryDiagnoser]
public class PoolAllocationBenchmark
{
    [Benchmark]
    public void Test()
    {
        for (var i = 0; i < 1_000_000; i++)
        {
            UsePool();
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.NoOptimization)]
    private static void UsePool()
    {
        var buffer  = ArrayPool<byte>.Shared.Rent(1024);
        Array.Clear(buffer);
    }
}
```

```
// * Summary *

BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4391/23H2/2023Update/SunValley3)
12th Gen Intel Core i9-12900HX, 1 CPU, 24 logical and 16 physical cores
.NET SDK 9.0.100-rc.2.24474.11
  [Host]     : .NET 8.0.10 (8.0.1024.46610), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.10 (8.0.1024.46610), X64 RyuJIT AVX2


| Method | Mean     | Error    | StdDev   | Gen0       | Allocated |
|------- |---------:|---------:|---------:|-----------:|----------:|
| Test   | 42.28 ms | 0.596 ms | 0.557 ms | 66750.0000 | 999.45 MB |
```